From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Tue, 30 Apr 2024 23:45:04 +0200
Subject: shell-app: Do not allocate memory during GSpawnChildSetupFunc

The child context setup function used as launch GSpawnChildSetupFunc
uses g_object_get that internally potentially allocates memory, making
it not async-signal-safe and so not something that is safe to use in
between fork and exec, so just use the simpler getter here.

Note that the current implementation of app_child_setup() is safe
already as sd_journal_stream_fd, dup2 and close are so.

Part-of: <https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/3293>
(cherry picked from commit dcb5956dea9c46720feaa1773ca9488327e34f51)
(cherry picked from commit f6c161ec6e3afe183c41f8ade950c832eaa47d19)
---
 src/shell-app.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/src/shell-app.c b/src/shell-app.c
index 3b5df0a..dcebc97 100644
--- a/src/shell-app.c
+++ b/src/shell-app.c
@@ -1236,9 +1236,8 @@ static void
 child_context_setup (gpointer user_data)
 {
   ShellGlobal *shell_global = user_data;
-  MetaContext *meta_context;
+  MetaContext *meta_context = shell_global_get_context (shell_global);
 
-  g_object_get (shell_global, "context", &meta_context, NULL);
   meta_context_restore_rlimit_nofile (meta_context, NULL);
 }
 
